### ISSUE TITLE
MessageBox convenience API?

### ISSUE THREAD

**Issue by iKarith** [marcusva/py-sdl2#170]
I just wrote this very simple program:

```c
#include "SDL2/SDL.h"

SDL_MessageBoxButtonData buttons[] = {
    { SDL_MESSAGEBOX_BUTTON_RETURNKEY_DEFAULT, 0, "Yes" },
    { SDL_MESSAGEBOX_BUTTON_ESCAPEKEY_DEFAULT, 1, "Cancel" },
};

SDL_MessageBoxColorScheme colorScheme = {
    {
        { 0x38, 0x36, 0x35 }, /* SDL_MESSAGEBOX_COLOR_BACKGROUND */
        { 0xd1, 0xcf, 0xcd }, /* SDL_MESSAGEBOX_COLOR_TEXT */
        { 0x8c, 0x87, 0x81 }, /* SDL_MESSAGEBOX_COLOR_BUTTON_BORDER */
        { 0x69, 0x66, 0x63 }, /* SDL_MESSAGEBOX_COLOR_BUTTON_BACKGROUND */
        { 0xcd, 0xca, 0x35 }  /* SDL_MESSAGEBOX_COLOR_BUTTON_SELECTED */
    }
};

SDL_MessageBoxData messageboxdata = {
    SDL_MESSAGEBOX_INFORMATION,
    NULL,                       /* No parent window */
    "Quit",                     /* Title */
    "Do you want to quit?",     /* Message */
    2,                          /* Number of buttons */
    buttons,                    /* The buttons themselves */
    &colorScheme                /* The color scheme */
};

int
main(int argc, char *argv[])
{
    int buttonid;
    if (SDL_ShowMessageBox(&messageboxdata, &buttonid) < 0) {
        SDL_Log("error displaying message box");
        return 1;
    }

    /* if (0 == buttonid) { do_quit(); } */
    /* What I do instead for this test is hokey AF */
    if (buttonid == 0) {
        SDL_Log("quitting as requested!");
    } else {
        SDL_Log("we shouldn't quit (but will anyway)");
	    SDL_Log("selection was %s",
                buttonid == -1 ? "to close the window" : buttons[buttonid].text);
    }
    return 0;
}
```

The Python needed to do the same with ctypes … was pain. Stuff like:

```python
ButtonData = sdl2.SDL_MessageBoxButtonData
ButtonDataArray = ButtonData * 2
MSGBOX_BUTTONS = ButtonDataArray(
    ButtonData(sdl2.SDL_MESSAGEBOX_BUTTON_ESCAPEKEY_DEFAULT, 0, b"Cancel"),
    ButtonData(sdl2.SDL_MESSAGEBOX_BUTTON_RETURNKEY_DEFAULT, 1, b"Yes"),
)
```

Had to assign ButtonData just to avoid having the line unavoidably longer than 79 characters long. (I'm not PEP8dantic, but … I try!) It's possible I don't have to do all of that—my ctypes-fu is not very strong, TBH. Something similar was required for the color scheme, and I had to ctypes.pointer() the resulting color scheme to add it to my sdl2.SDL_MessageBoxData arguments.

The color scheme above in the C program mimics that used on X11 for the SimpleMessageBox. I tend to prefer working with hex values for colors, it's just a habit from HTML and it conveniently makes the code line up nicely. And of course I just realized I could have set it to NULL and bypassed reading the source to learn what the default colors were.

Still, could we have some sdl.ext.msgbox goodness to clean that up a bit?

**Comment by a-hurst** [comment]
@iKarith This is extremely valuable feedback, thank you! With all the different APIs within SDL2 it's hard to test whether they all work nicely or correctly (especially the ones that require a GUI or external hardware, since that makes unit testing via CI difficult), so it's handy to know how well a specific part is working with the ctypes wrappers.

I think the messagebox API is a useful one to have in pysdl2, so if it's this unfriendly to use with the wrapped ctypes functions I think a .ext module is warranted. Would you be willing to share your full Python version of the above C code? that would save me some time getting started on a wrapper and help me better assess the situation

**Comment by iKarith** [comment]
I unfortunately didn't write a full python version of the above code because shortly after submitting that, I ripped out the partial code (it was only intended to be temporary anyway) and implemented an advanced debugging solution involving print(). 🤣 I actually learned that there are some defaults if I'd chosen to use them. My C code does things the hard way. This really is intended to be used for fire and forget debugging purposes.
